File: golang/producer.go
package main

import (
	"crypto/tls"
	"github.com/IBM/sarama"
	"gitlab.vsk.ru/marlin/examples/marlin-kafka-examples/golang/provider"
	"log"
	"strings"
)

func main() {
	bootstrapServers := "kafka1.example.org:443,kafka2.example.org:443"
	clientID := "oauth-client-id"
	clientSecret := "oauth-client-secret"
	tokenEndpoint := "https://keycloak.example.org/auth/realms/kafka-authz/protocol/openid-connect/token"
	topic := "example"

	c := sarama.NewConfig()
	c.Net.TLS.Enable = true
	c.Net.TLS.Config = &tls.Config{
		InsecureSkipVerify: true,
	}
	c.Net.SASL.Enable = true
	c.Net.SASL.Mechanism = sarama.SASLTypeOAuth
	c.Net.SASL.TokenProvider = provider.NewTokenProvider(clientID, clientSecret, tokenEndpoint)
	c.Producer.RequiredAcks = sarama.WaitForLocal
	c.Producer.Return.Successes = true

	producer, err := sarama.NewSyncProducer(strings.Split(bootstrapServers, ","), c)
	if err != nil {
		panic(err)
	}

	msg := &sarama.ProducerMessage{
		Topic: topic,
		Value: sarama.StringEncoder("A test message"),
	}
	partition, offset, err := producer.SendMessage(msg)
	if err != nil {
		panic(err)
	}
	log.Printf("Message sent: partition=%d, offset=%d\n", partition, offset)
}

File: nodejs/producer.js
const {Kafka} = require('kafkajs');
const axios= require('axios');

const brokers = 'kafka1.example.org:443,kafka2.example.org:443'
const clientId = 'oauth-client-id';
const clientSecret = 'oauth-client-secret';
const tokenEndpoint = 'https://keycloak.example.org/auth/realms/kafka-authz/protocol/openid-connect/token'
const topic = 'example';

async function getAccessToken() {
    return new Promise((resolve, reject) => {
        axios.post(tokenEndpoint, new URLSearchParams({
            grant_type: 'client_credentials',
            client_id: clientId,
            client_secret: clientSecret
        })).then(res => {
            resolve(res.data.access_token);
        }, err => {
            console.log('Error: ', err.message);
            reject(err);
        });
    });
}

async function main() {
    // Set up the Kafka client
    const kafka = new Kafka({
        clientId: 'marlin-kafka-example-app',
        brokers: brokers.split(','),
        ssl: true,
        sasl: {
            mechanism: 'oauthbearer',
            oauthBearerProvider: async () => {
                return {value: await getAccessToken()}
            }
        },
    });

    const producer = kafka.producer();
    await producer.connect();

    const metadata = await producer.send({
        topic: topic,
        messages: [{key: "example_key", value: "Example message"}],
    });

    console.log(`Message was sent to the partition ${metadata[0].partition}. New offset is ${metadata[0].baseOffset}`);

    // Close the producer
    await producer.disconnect();
}

main().catch(console.error);
